File: types/asset.ts
export interface AssetDetail {
  ticker: string;
  name: string;
  sector: string;
  price: number;
  change: number;
  changePercent: number;
  volume: number;
  marketCap: number;
  signal: "bullish" | "bearish" | "neutral";
  signalScore: number;
  priceHistory: PricePoint[];
}

export interface PricePoint {
  date: string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}

export interface TechnicalIndicator {
  name: string;
  value: number;
  signal: "bullish" | "bearish" | "neutral";
  description: string;
}

export interface AssetTechnicalResponse {
  ticker: string;
  indicators: TechnicalIndicator[];
  ema: { period: number; value: number }[];
  rsi: number;
  macd: { macd: number; signal: number; histogram: number };
  atr: number;
  volatility: number;
}

export interface ForecastModelOutput {
  modelName: string;
  status: "live" | "simulated" | "coming_soon";
  directionProbability: number;
  predictedDirection: "up" | "down";
  expectedReturn: number;
  confidence: number;
  explanation: string;
}

export interface AssetForecastResponse {
  ticker: string;
  models: ForecastModelOutput[];
  bullishFactors: string[];
  bearishFactors: string[];
  riskLevel: "low" | "medium" | "high";
  aiSummary: string;
}

File: backend/app/schemas/asset.py
from __future__ import annotations

from pydantic import BaseModel
from typing import Optional


class PricePoint(BaseModel):
    date: str
    open: float
    high: float
    low: float
    close: float
    volume: int


class AssetDetailResponse(BaseModel):
    ticker: str
    name: str
    sector: str
    price: float
    change: float
    changePercent: float
    volume: int
    marketCap: int
    signal: str
    signalScore: float
    priceHistory: list[PricePoint]


class TechnicalIndicator(BaseModel):
    name: str
    value: float
    signal: str
    description: str


class EMAValue(BaseModel):
    period: int
    value: float


class MACDValue(BaseModel):
    macd: float
    signal: float
    histogram: float


class AssetTechnicalResponse(BaseModel):
    ticker: str
    indicators: list[TechnicalIndicator]
    ema: list[EMAValue]
    rsi: float
    macd: MACDValue
    atr: float
    volatility: float


class ForecastModelOutput(BaseModel):
    modelName: str
    status: str  # "live" | "simulated" | "coming_soon"
    directionProbability: float
    predictedDirection: str
    expectedReturn: float
    confidence: float
    explanation: str


class AssetForecastResponse(BaseModel):
    ticker: str
    models: list[ForecastModelOutput]
    bullishFactors: list[str]
    bearishFactors: list[str]
    riskLevel: str
    aiSummary: str
